third_party/vcluster/vcluster.go

```diff
@@ -0,0 +1,324 @@
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package vcluster
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+
+	"github.com/vladimirvivien/gexe"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/rest"
+	log "k8s.io/klog/v2"
+	"sigs.k8s.io/e2e-framework/klient"
+	"sigs.k8s.io/e2e-framework/klient/conf"
+	"sigs.k8s.io/e2e-framework/klient/k8s/resources"
+	"sigs.k8s.io/e2e-framework/klient/wait"
+	"sigs.k8s.io/e2e-framework/klient/wait/conditions"
+	"sigs.k8s.io/e2e-framework/support"
+	"sigs.k8s.io/e2e-framework/support/utils"
+	"sigs.k8s.io/yaml"
+)
+
+var vclusterVersion = "v0.20.0"
+
+type Cluster struct {
+	path            string
+	name            string
+	kubecfgFile     string // kubeconfig file for the vcluster
+	version         string
+	namespace       string // namespace to create the vcluster in
+	hostKubeCfg     string // kubeconfig file for the host cluster
+	hostKubeContext string // kubeconfig context for the host cluster
+	rc              *rest.Config
+}
+
+// Enforce Type check always to avoid future breaks
+var _ support.E2EClusterProvider = &Cluster{}
+
+func NewCluster(name string) *Cluster {
+	return &Cluster{name: name}
+}
+
+func NewProvider() support.E2EClusterProvider {
+	return &Cluster{}
+}
+
+func WithPath(path string) support.ClusterOpts {
+	return func(c support.E2EClusterProvider) {
+		v, ok := c.(*Cluster)
+		if ok {
+			v.path = path
+		}
+	}
+}
+
+func WithNamespace(ns string) support.ClusterOpts {
+	return func(c support.E2EClusterProvider) {
+		v, ok := c.(*Cluster)
+		if ok {
+			v.namespace = ns
+		}
+	}
+}
+
+func WithHostKubeConfig(kubeconfig string) support.ClusterOpts {
+	return func(c support.E2EClusterProvider) {
+		v, ok := c.(*Cluster)
+		if ok {
+			v.hostKubeCfg = kubeconfig
+		}
+	}
+}
+
+func WithHostKubeContext(context string) support.ClusterOpts {
+	return func(c support.E2EClusterProvider) {
+		v, ok := c.(*Cluster)
+		if ok {
+			v.hostKubeContext = context
+		}
+	}
+}
+
+func (c *Cluster) WithName(name string) support.E2EClusterProvider {
+	c.name = name
+	return c
+}
+
+func (c *Cluster) WithVersion(version string) support.E2EClusterProvider {
+	c.version = version
+	return c
+}
+
+func (c *Cluster) WithPath(path string) support.E2EClusterProvider {
+	c.path = path
+	return c
+}
+func (c *Cluster) WithOpts(opts ...support.ClusterOpts) support.E2EClusterProvider {
+	for _, opt := range opts {
+		opt(c)
+	}
+	return c
+}
+func (c *Cluster) SetDefaults() support.E2EClusterProvider {
+	if c.path == "" {
+		c.path = "vcluster"
+	}
+	return c
+}
+func (c *Cluster) Create(ctx context.Context, args ...string) (string, error) {
```

Comment:
@crandles does this vcluster create function will only work for kind "real" clusters (i don't know much about vclusters). If that is the case, you may wanna consider making that clear in naming (i.e. third_party/kind_vcluster).

Reply:
No, this doesn't require kind, this should work in any host k8s cluster.

I set it up via kind in the example so that it could presumably execute successfully (doubting there's a real cluster to access otherwise within CI).